wtxmgr/db.go

```diff
@@ -0,0 +1,1311 @@
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+package wtxmgr
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// Naming
+//
+// The following variables are commonly used in this file and given
+// reserved names:
+//
+//   ns: The namespace bucket for this package
+//   b:  The primary bucket being operated on
+//   k:  A single bucket key
+//   v:  A single bucket value
+//   c:  A bucket cursor
+//   ck: The current cursor key
+//   cv: The current cursor value
+//
+// Functions use the naming scheme `Op[Raw]Type`, which performs the
+// operation `Op` on the type `Type`, optionally dealing with raw keys
+// and values if `Raw` is used.  The following operations are used:
+//
+//   key:    return a db key for some data
+//   value:  return a db value for some data
+//   put:    insert or replace a value into a bucket
+//   fetch:  read and return a value
+//   read:   read a value into an out parameter
+//   exists: return the raw (nil if not found) value for some data
+//   delete: remove a k/v pair
+//
+// Other operations which are specific to the types being operated on
+// should be explained in a comment.
+
+// Big endian is the preferred byte order, due to cursor scans over integer
+// keys iterating in order.
+var byteOrder = binary.BigEndian
+
+const (
+	// LatestTxStoreVersion is the most recent tx store version.
+	LatestTxStoreVersion = 1
+)
+
+// Bucket names
+var (
+	bucketBlocks         = []byte("b")
+	bucketTxRecords      = []byte("t")
+	bucketCredits        = []byte("c")
+	bucketUnspent        = []byte("u")
+	bucketDebits         = []byte("d")
+	bucketUnmined        = []byte("m")
+	bucketUnminedCredits = []byte("mc")
+	bucketUnminedInputs  = []byte("mi")
+)
+
+// Root (namespace) bucket keys
+var (
+	rootCreateDate   = []byte("date")
+	rootVersion      = []byte("vers")
+	rootMinedBalance = []byte("bal")
+)
+
+// The root bucket's mined balance k/v pair records the total balance for all
+// unspent credits from mined transactions.  This includes immature outputs, and
+// outputs spent by mempool transactions, which must be considered when
+// returning the actual balance for a given number of block confirmations.  The
+// value is the amount serialized as a uint64.
+
+func fetchMinedBalance(ns walletdb.Bucket) (btcutil.Amount, error) {
+	v := ns.Get(rootMinedBalance)
+	if len(v) != 8 {
+		str := fmt.Sprintf("balance: short read (expected 8 bytes, "+
+			"read %v)", len(v))
+		return 0, storeError(ErrData, str, nil)
+	}
+	return btcutil.Amount(byteOrder.Uint64(v)), nil
+}
+
+func putMinedBalance(ns walletdb.Bucket, amt btcutil.Amount) error {
+	v := make([]byte, 8)
+	byteOrder.PutUint64(v, uint64(amt))
```

Comment:
Maybe.  The only reason I'd see this mattering is if you wanted to optimize the call to `binary.LittleEndian.PutUintXX` on little endian hosts, but it doesn't do that at the moment (at least in the source code, I haven't checked compiler output).